src/Core/Resolvers/CosmosQueryEngine.cs

```diff
@@ -80,6 +90,48 @@ public class CosmosQueryEngine : IQueryEngine
                 queryRequestOptions.PartitionKey = new PartitionKey(partitionKeyValue);
             }
 
+            JObject executeQueryResult = null;
+
+            if (runtimeConfig.CanUseCache() && runtimeConfig.Entities[structure.EntityName].IsCachingEnabled)
+            {
+                StringBuilder dataSourceKey = new(dataSourceName);
+
+                // to support caching for paginated query adding continuation token in the datasource
+                if (structure.IsPaginated)
+                {
+                    dataSourceKey.Append(":");
+                    dataSourceKey.Append(structure.IsPaginated);
+                    dataSourceKey.Append(":");
+                    dataSourceKey.Append(structure.Continuation);
+                }
+
+                DatabaseQueryMetadata queryMetadata = new(queryText: queryString, dataSource: dataSourceKey.ToString(), queryParameters: structure.Parameters);
```

Comment:
`why is pagination metadata used for DatabaseQueryMetadata.dataSource as opposed to queryText?`
Cosmos DB supports paginated queries, where the query remains constant while the continuation token changes. Cosmos DB utilizes this continuation token to retrieve the next set of records. Due to this functionality, relying solely on the query text as the key is undesirable, as it would result in receiving the same response for each page.
Here, we anticipate that the user will provide the same token again. Given that the initial cached response will return the cached token value, subsequent calls will be handled by the cache. If the user attempts to tamper with the continuation token, cached data will not be returned, and later, when the SDK makes a call, it will fail at the Cosmos DB end as well.

` Also should the Continuation token be used in its raw form here within dataSourceKey? what about added it as a queryParameter to avoid "sql" injection?`
We are appending the continuation token to the dataSourceKey, which will serve as our cache key. At no point are we executing any operations on this text. Furthermore, any changes or injections in the continuation key are already handled at the Cosmos DB end. Therefore, we should be in good shape.

Reply:
Right. So, this will be the sequence.
First query -> SDK -> Cache
Second query with continuation token -> SDK -> Cache
End

Now, if user again request make the same first query, we will return cached data which will have the token for second query. And the response of which is also cached.